experimental/ast/path.go

```diff
@@ -266,20 +251,87 @@ func (p PathComponent) AsExtension() Path {
 // AsIdent returns the single identifier that makes up this path component, if
 // it is not an extension path component.
 //
-// May be nil, in the case of e.g. the second component of foo..bar
+// May be nil, in the case of e.g. the second component of foo..bar.
 func (p PathComponent) AsIdent() token.Token {
 	return p.name.In(p.Context())
 }
 
+// rawPath is the raw contents of a Path without its Context.
+//
+// This has one of the following configurations.
+//
+//  1. Two zero tokens. This is the nil path.
+//
+//  2. Two natural tokens. This means the path is all tokens between them including
+//     the end-point
+//
+//  3. A single synthetic token and a nil token. If this token has children, those are
+//     the path components. Otherwise, the token itself is the sole token.
+//
+// The case Start < 0 && End != 0 is reserved for use by pathLike below.
+type rawPath struct {
+	Start, End token.ID
+}
+
 // Wrap wraps this rawPath with a context to present to the user.
 func (p rawPath) With(c Context) Path {
-	if p[0].Nil() {
-		return PathNil
+	if p.Start.Nil() {
+		return Path{}
 	}
 
-	if p[1].Nil() {
+	if p.End.Nil() {
 		panic(fmt.Sprintf("protocompile/ast: invalid ast.Path representation %v; this is a bug in protocompile", p))
 	}
 
 	return Path{internal.NewWith(c), p}
 }
+
+// pathLike is the raw representation of a type or expression.
+//
+// The vast, vast majority of types and expressions are paths. To avoid needing
+// to waste space for such types, we use the following encoding for rawType.
+//
+// First, note that if the first half of a rawPath is negative, the other
+// must be zero. Thus, if the first "token" of the rawPath is negative and
+// the second is not, the first is ^typeKind and the second is an index
+// into a table in a Context. Otherwise, it's a path type. This logic is
+// implemented in With().
+type pathLike[Kind ~int8] rawPath
+
+// wrapInPathLike wraps a integer-like value in a rawPathOrPointer.
+//
+// If either of kind or value are zero, both must be.
+func wrapPathLike[Value ~int32 | ~uint32, Kind ~int8](kind Kind, value Value) pathLike[Kind] {
+	if kind != 0 && value == 0 {
+		panic(fmt.Sprintf("protocompile/ast: invalid rawPathOrPointer representation: %v, %v", kind, value))
+	}
+
+	return pathLike[Kind]{^token.ID(kind), token.ID(value)}
+}
+
+// unwrapPathLike unwraps a pointer previously wrapped with wrapPathLike.
+//
+// Returns zero if this pathLike contains the wrong kind.
+func unwrapPathLike[Value ~int32 | ~uint32, Kind ~int8](want Kind, p pathLike[Kind]) Value {
+	if got, ok := p.kind(); !ok || got != want {
+		return 0
+	}
+
+	return Value(p.End)
```

Comment:
Maybe a comment to clarify that if this is a path where the start happens to equal the wanted kind, this will always return zero (per invariants of a path and how kind is encoded).

Reply:
(Sorry -- didn't mean to send this as a one-off remark. I meant to add to a set of review comments. Feel free to ignore this until I send the other comments.)